### PR TITLE
[Crypto] improve shuffle test

### PR DESCRIPTION
**Pull request by tarakby**
Improve a statistical test that evaluates the uniformity of a random shuffling.
The previous and new versions are both sanity checks of uniformity that can detect bugs in the implementation, but are not extended statistical tests. 

The previous version chooses a random element of the original array, and tracks the distribution of indices where this element falls in the shuffled array. If the shuffling is uniform, the distribution of indices is uniform too. Uniformity is simply evaluated by comparing the standard deviation to a low threshold.  
The new version uses a known bijection between the set of all permutations (shuffles) and a set of integers, using the Lehmer code the factorial number radix. If the shuffling is uniform, the distribution of the permutation encodings is uniform too. Uniformity is still evaluated using the same basic standard deviation.

The new version evaluates the permutation as a whole instead of isolating one element only. 
The drawback is that the evaluated distribution in the new version is of size `n!` (compared to `n` in the previous version), so the tested shuffled array has to remain small.

Side change:
Export the `EncodePermutation` function so that other permutation tests can use the tool.

### CODE REVIEW

File: crypto/random/rand_utils.go [hunk 1]
```diff
@@ -49,3 +53,27 @@ func EvaluateDistributionUniformity(t *testing.T, distribution []float64) {
 	mean := stat.Mean(distribution, nil)
 	assert.Greater(t, tolerance*mean, stdev, fmt.Sprintf("basic randomness test failed: n: %d, stdev: %v, mean: %v", len(distribution), stdev, mean))
 }
+
+// computes a bijection from the set of all permutations
+// into the the set [0, n!-1] (where `n` is the size of input `perm`).
+// input `perm` is assumed to be a correct permutation of the set [0,n-1]
+// (not checked in this function).
+func EncodePermutation(perm []int) int {
+	r := make([]int, len(perm))
```
Comment (jordanschalm) [inline crypto/random/rand_utils.go:62]:
```suggestion
	// We first construct r, which contains elements with the following constraints:
	//   r = { r(0) ∈ [0,n-1], r(1) ∈ [0,n-2], ... , r(n-1) ∈ [0,1], r(n) ∈ [0] }
	r := make([]int, len(perm))
```

File: crypto/random/rand_utils.go [hunk 2]
```diff
@@ -49,3 +53,27 @@ func EvaluateDistributionUniformity(t *testing.T, distribution []float64) {
...
+	// generate Lehmer code
+	for i, x := range perm {
+		for _, y := range perm[i+1:] {
+			if y < x {
+				r[i]++
+			}
+		}
+	}
+	// Convert to an integer following the factorial number system
+	m := 0
```
Comment (jordanschalm) [inline crypto/random/rand_utils.go:72]:
```suggestion
	// We construct `m` by taking the sum of all elements of `r`, each multiplied by the factorial of their index:
	// m = ∑ r(i) * i!, i ∈ [0,n-1]
	m := 0
```

File: crypto/random/rand_test.go [hunk 3]
```diff
@@ -228,52 +226,47 @@ func TestShuffle(t *testing.T) {
...
+				permEncoding := func() (uint64, error) {
+					err = rng.Shuffle(listSize, func(i, j int) {
+						list[i], list[j] = list[j], list[i]
+					})
+					return uint64(EncodePermutation(list)), err
+				}
+				BasicDistributionTest(t, factN, 1, permEncoding)
+			})
+
+			t.Run("shuffle a same permutation", func(t *testing.T) {
```
Comment (jordanschalm) [inline crypto/random/rand_test.go:255]: I don't understand why we have this test, which looks identical to the previous one.

Comment (tarakby) [reply crypto/random/rand_test.go:255]: - the first test shuffles an array multiple times. The resulting array from the previous shuffle is shuffled again. The distribution is computed based on all intermediate arrays from each loop.
- the second test shuffles the same array [0,1, .. , n-1] multiple times. The resulting array after each loop is used for the distribution, but it is then ignored. The next loop shuffle is computed using the same starting array. 

The shuffling should be uniform in both scenarios. Some buggy implementations may be uniform in one scenario and not the other.

File: crypto/random/rand_utils.go (hunk 1 above)
Comment (tarakby) [reply crypto/random/rand_utils.go:62]: The suggested comment is the start of explaining what a Lehmer code is. However it doesn't fully explain how `r` is obtained, or why there is a bijection between all permutations and the set `[0,n-1] x [0,n-2] x ... x [0,1] x [0]`, so IMO it doesn't add much information. 

In the original comment I only mentioned "compute Lehmer code", maybe I can add a link to what Lehmer code is (that contains all the details).

File: crypto/random/rand_utils.go (hunk 2 above)
Comment (tarakby) [reply crypto/random/rand_utils.go:72]: This comment is more explanatory but it doesn't explain why it makes sense to compute `m` (why there is a bijection between `[0,n-1] x [0,n-2] x ... x [0,1] x [0]` and `[0, n! -1] ). 
I suggest to add a link to the full explanation instead.

Comment (tarakby) [reply crypto/random/rand_utils.go:74]: https://github.com/onflow/flow-go/pull/4844/commits/03b9532779fe2e8c9504bc3472e9131dfc7d0c1f

File: crypto/random/rand_utils.go (hunk 1 above)
Comment (tarakby) [reply crypto/random/rand_utils.go:62]: https://github.com/onflow/flow-go/pull/4844/commits/03b9532779fe2e8c9504bc3472e9131dfc7d0c1f

**Review by jordanschalm — APPROVED** [review]

**Review by sisyphusSmiling — APPROVED** [review]

**Pull request merged**